implementations/micrometer-registry-statsd/src/test/java/io/micrometer/statsd/StatsdMeterRegistryPublishTest.java

```diff
@@ -0,0 +1,234 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.micrometer.statsd;
+
+import io.micrometer.core.Issue;
+import io.micrometer.core.instrument.Clock;
+import io.micrometer.core.instrument.Counter;
+import io.netty.handler.logging.LogLevel;
+import org.jetbrains.annotations.NotNull;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.params.ParameterizedTest;
+import org.junit.jupiter.params.provider.EnumSource;
+import reactor.core.publisher.Flux;
+import reactor.netty.DisposableChannel;
+import reactor.netty.tcp.TcpServer;
+import reactor.netty.udp.UdpServer;
+
+import java.time.Duration;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.TimeUnit;
+import java.util.stream.IntStream;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.awaitility.Awaitility.await;
+
+/**
+ * Tests {@link StatsdMeterRegistry} metrics publishing functionality.
+ */
+class StatsdMeterRegistryPublishTest {
+
+    StatsdMeterRegistry meterRegistry;
+    CountDownLatch serverLatch;
+
+    @ParameterizedTest
+    @EnumSource(StatsdProtocol.class)
+    void receiveMetricsSuccessfully(StatsdProtocol protocol) throws InterruptedException {
+        serverLatch = new CountDownLatch(3);
+        DisposableChannel server = startServer(protocol, 0);
+
+        final int port = server.address().getPort();
+
+        meterRegistry = new StatsdMeterRegistry(getUnbufferedConfig(protocol, port), Clock.SYSTEM);
+        startRegistryAndWaitForClient();
+        Counter counter = Counter.builder("my.counter").register(meterRegistry);
+        counter.increment();
+        counter.increment();
+        counter.increment();
+        assertThat(serverLatch.await(3, TimeUnit.SECONDS)).isTrue();
+        meterRegistry.close();
+
+        server.disposeNow();
+    }
+
+    @ParameterizedTest
+    @EnumSource(StatsdProtocol.class)
+    void resumeSendingMetrics_whenServerIntermittentlyFails(StatsdProtocol protocol) throws InterruptedException {
+        serverLatch = new CountDownLatch(1);
+        DisposableChannel server = startServer(protocol, 0);
+
+        final int port = server.address().getPort();
+
+        meterRegistry = new StatsdMeterRegistry(getUnbufferedConfig(protocol, port), Clock.SYSTEM);
+        startRegistryAndWaitForClient();
+        Counter counter = Counter.builder("my.counter").register(meterRegistry);
+        counter.increment(1);
+        assertThat(serverLatch.await(5, TimeUnit.SECONDS)).isTrue();
+        // stop server
+        server.disposeNow();
+        serverLatch = new CountDownLatch(3);
+        // client will try to send but server is down
+        IntStream.range(2, 100).forEach(counter::increment);
+        // TODO wait until processor queue is drained...
+        // start server
+        server = startServer(protocol, port);
+        assertThat(serverLatch.getCount()).isEqualTo(3);
```

Comment:
This is consistently failing in the TCP case because all increments while the server is down are received and processed by the server when it starts up. Is this expected?

Reply:
I tried to debug to find why, but it seems that for some reason the server isn't properly shutting down and keeps receiving writes from the registry :( pinging @violetagg for investigation.

Reply:
Violeta's suggestion in another comment worked. TCP tests are passing consistently now. Thanks.